modules/tensor_mechanics/src/interfacekernels/CZMInterfaceKernelBase.C

```diff
@@ -0,0 +1,108 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "CZMInterfaceKernelBase.h"
+
+InputParameters
+CZMInterfaceKernelBase::validParams()
+{
+  InputParameters params = InterfaceKernel::validParams();
+  params.addRequiredParam<unsigned int>("component",
+                                        "the component of the "
+                                        "displacement vector this kernel is working on:"
+                                        " component == 0, ==> X"
+                                        " component == 1, ==> Y"
+                                        " component == 2, ==> Z");
+  params.set<bool>("_use_undisplaced_reference_points") = true;
+  params.suppressParameter<bool>("use_displaced_mesh");
+  params.addRequiredCoupledVar("displacements", "the string containing displacement variables");
+  params.addParam<std::string>("base_name", "Material property base name");
+  params.set<std::string>("traction_global_name") = "traction_global";
+
+  return params;
+}
+
+CZMInterfaceKernelBase::CZMInterfaceKernelBase(const InputParameters & parameters)
+  : JvarMapKernelInterface<InterfaceKernel>(parameters),
+    _base_name(isParamValid("base_name") && !getParam<std::string>("base_name").empty()
+                   ? getParam<std::string>("base_name") + "_"
+                   : ""),
+    _component(getParam<unsigned int>("component")),
+    _ndisp(coupledComponents("displacements")),
+    _disp_var(_ndisp),
+    _disp_neighbor_var(_ndisp),
+    _vars(_ndisp),
+    _traction_global(getMaterialPropertyByName<RealVectorValue>(
+        _base_name + getParam<std::string>("traction_global_name"))),
+    _dtraction_djump_global(
+        getMaterialPropertyByName<RankTwoTensor>(_base_name + "dtraction_djump_global"))
+{
+  // Enforce consistency
+  if (_ndisp != _mesh.dimension())
+    paramError("displacements", "Number of displacements must match problem dimension.");
+
+  if (_ndisp > 3 || _ndisp < 1)
+    mooseError("the CZM material requires 1, 2 or 3 displacement variables");
+}
+
+void
+CZMInterfaceKernelBase::initialSetup()
+{
+  for (unsigned int i = 0; i < _ndisp; ++i)
+  {
+    _disp_var[i] = coupled("displacements", i);
+    _disp_neighbor_var[i] = coupled("displacements", i);
+    _vars[i] = getVar("displacements", i);
+  }
+}
```

Comment:
why not in the constructor?

Reply:
At construction time the constant reference might not have been set yet. Also this appears to be a cleaner implementation, though it would require explicitly calling it in derived classes.